harper-core/src/linting/duplicate_punctuation.rs

```diff
@@ -0,0 +1,197 @@
+                    i = end + 1;
+                    continue;
+                }
+            }
+            i += 1;
+        }
+        lints
+    }
+
+    fn description(&self) -> &str {
+        "Flags duplicate or mixed punctuation sequences."
+    }
+}
+
+fn is_interest(p: Punctuation) -> bool {
+    matches!(
+        p,
+        Punctuation::Comma
+            | Punctuation::Semicolon
+            | Punctuation::Colon
+            | Punctuation::Bang
+            | Punctuation::Question
+            | Punctuation::ForwardSlash
+            | Punctuation::Ampersand
+    )
+}
+
+fn in_same_group(first: Punctuation, next: Punctuation) -> bool {
+    match first {
+        Punctuation::Bang | Punctuation::Question => {
+            matches!(next, Punctuation::Bang | Punctuation::Question)
+        }
+        _ => first == next,
+    }
+}
+
+fn build_suggestions(seq: &[Punctuation]) -> Vec<Suggestion> {
+    use Punctuation::*;
+    let mut opts = match seq[0] {
+        Bang | Question => vec![Bang, Question],
+        other => vec![other],
+    };
+    opts.dedup();
+    opts.into_iter()
+        .map(|p| Suggestion::ReplaceWith(vec![punct_char(p)]))
+        .collect()
+}
+
+fn punct_char(p: Punctuation) -> char {
+    use Punctuation::*;
+    match p {
+        Comma => ',',
+        Semicolon => ';',
+        Colon => ':',
+        Bang => '!',
+        Question => '?',
+        ForwardSlash => '/',
+        Ampersand => '&',
+        _ => unreachable!(),
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::linting::tests::assert_lint_count;
+
+    use super::DuplicatePunctuation;
+
+    #[test]
+    fn flags_double_comma() {
+        assert_lint_count("Wait,, what happened?", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_double_semicolon() {
+        assert_lint_count("He hesitated;; then spoke.", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_double_colon() {
+        assert_lint_count("Choices:: A or B.", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_double_bang() {
+        assert_lint_count("Stop!!", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_double_question() {
+        assert_lint_count("Really??", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_mixed_qbang_pair() {
+        assert_lint_count("What?!", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_triple_bang() {
+        assert_lint_count("No!!! Absolutely not.", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_q_bang_bang() {
+        assert_lint_count("Really?!!", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_double_slash() {
+        assert_lint_count("This // is a typo.", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_triple_slash() {
+        assert_lint_count("Path error: ///tmp.", DuplicatePunctuation, 1);
+    }
+
+    #[test]
+    fn flags_triple_question() {
+        assert_lint_count("Why???", DuplicatePunctuation, 1);
```

Comment:
This one could be used in a dialog to emphasis the surprise. I'm not saying it's valid in English, but it's common on the internet